File: python/secid_client.py
#!/usr/bin/env python3
"""SecID client — resolve security identifiers to URLs.

Single file, zero dependencies (stdlib only). Copy and use.

SecID is a universal grammar for security knowledge:
    secid:type/namespace/name[@version]#subpath

API: GET https://secid.cloudsecurityalliance.org/api/v1/resolve?secid={encoded}

IMPORTANT: The # character in SecID strings must be encoded as %23 in the
URL query parameter. This is the #1 failure mode for new clients.

Usage as library:
    from secid_client import SecIDClient
    client = SecIDClient()
    response = client.resolve("secid:advisory/mitre.org/cve#CVE-2021-44228")
    print(response.best_url)

Usage as CLI:
    secid "secid:advisory/mitre.org/cve#CVE-2021-44228"
    secid --json "secid:advisory/mitre.org/cve#CVE-2021-44228"
"""

from __future__ import annotations

__version__ = "0.1.0"

import http.client
import json
import math
import socket
import sys
import urllib.error
import urllib.parse
import urllib.request
from dataclasses import dataclass, field
from typing import Any

DEFAULT_BASE_URL = "https://secid.cloudsecurityalliance.org"
DEFAULT_TIMEOUT = 30  # seconds
MAX_RESPONSE_BYTES = 10 * 1024 * 1024  # 10 MB


@dataclass
class SecIDResponse:
    """Response from the SecID resolve API.

    Attributes:
        secid_query: The query string echoed back (decoded form).
        status: One of: found, corrected, related, not_found, error.
        results: List of result dicts — either resolution or registry type.
            Resolution results have: secid, weight, url, and optional
            content_type, parsability ("structured"/"scraped"), schema
            (SecID reference), parsing_instructions (SecID reference),
            auth (free-text).
        message: Guidance text on not_found/error, None otherwise.
    """

    secid_query: str
    status: str
    results: list[dict[str, Any]] = field(default_factory=list)
    message: str | None = None

    @property
    def best_url(self) -> str | None:
        """Highest-weight valid URL from resolution results, or None.

        Only absolute http(s) URLs count (see _validate_url). If the top
        result carries a hostile or malformed URL, the next valid one is
        returned instead of None.
        """
        resolved = self.resolution_results
        return resolved[0]["url"] if resolved else None

    @property
    def was_corrected(self) -> bool:
        """True if the server corrected the input."""
        return self.status == "corrected"

    @property
    def resolution_results(self) -> list[dict[str, Any]]:
        """Only results with a numeric weight and a valid http(s) url, sorted
        by weight descending. Entries with a missing, null, or non-numeric
        weight, or a url that fails _validate_url (javascript:, data:,
        relative, ...), are left out."""
        return sorted(
            [r for r in self.results
             if isinstance(r, dict) and _is_weight(r.get("weight"))
             and _validate_url(r.get("url")) is not None],
            key=lambda r: r["weight"],
            reverse=True,
        )

    @property
    def registry_results(self) -> list[dict[str, Any]]:
        """Only results with data (registry/browsing info)."""
        return [r for r in self.results if isinstance(r, dict) and "data" in r]


def _is_weight(value: Any) -> bool:
    """True for a finite int/float. bool is excluded (it is an int subclass)."""
    return (isinstance(value, (int, float)) and not isinstance(value, bool)
            and math.isfinite(value))


# The resolver response is untrusted (a hostile, federated, or MITM'd resolver
# is in scope). Only absolute http(s) URLs may be surfaced. The check is done
# by hand rather than with urlparse so that all three reference clients apply
# exactly the same rule; parsers disagree on inputs such as "http:evil",
# "http:///evil" and "http:\\evil" (a WHATWG parser accepts all three).
ALLOWED_URL_SCHEMES = frozenset({"https", "http"})


def _validate_url(url: Any) -> str | None:
    """Return url only if it is an absolute http(s) URL with a host, else None.

    Rules (identical in the Python, TypeScript and Go clients):
      1. a non-empty string with no ASCII control characters or spaces;
      2. begins with "http://" or "https://" (scheme case-insensitive);
      3. the authority (up to the first "/", "?", "#" or backslash), minus
         any "userinfo@", is non-empty and does not start with ":".
    """
    if not isinstance(url, str) or not url:
        return None
    if any(ord(ch) <= 0x20 or ord(ch) == 0x7F for ch in url):
        return None
    scheme, sep, rest = url.partition("://")
    if not sep or scheme.lower() not in ALLOWED_URL_SCHEMES:
        return None
    end = len(rest)
    for delim in "/?#\\":
        i = rest.find(delim)
        if i != -1:
            end = min(end, i)
    host = rest[:end].rpartition("@")[2]
    if not host or host.startswith(":"):
        return None
    return url


def _sanitize_terminal(text: Any) -> str:
    """Strip C0/C1 control chars (incl. ESC) from server-controlled text before
    printing to a terminal — prevents ANSI/escape-sequence injection.

    Accepts any value: a hostile resolver can send a number or object where a
    string is expected, and printing it must not crash the CLI."""
    if not isinstance(text, str):
        text = "" if text is None else str(text)
    return "".join(
        ch for ch in text
        if not (ord(ch) < 0x20 or ord(ch) == 0x7F or 0x80 <= ord(ch) <= 0x9F)
    )


_INVALID = object()

# socket.timeout is an alias of TimeoutError from 3.10; on 3.9 it is a
# separate OSError subclass.
_TIMEOUTS = (TimeoutError, socket.timeout)


def _parse_json(body: bytes) -> Any:
    """Decode a response body as JSON, returning _INVALID on any failure."""
    try:
        return json.loads(body.decode("utf-8"))
    except (ValueError, UnicodeDecodeError, RecursionError):
        return _INVALID


def _envelope(data: dict[str, Any], secid: str) -> SecIDResponse:
    """Build a SecIDResponse from an untrusted JSON object, type-checking each
    field. Wrong-typed fields fall back to defaults; non-object results are
    dropped so the result helpers never see them."""
    query = data.get("secid_query")
    status = data.get("status")
    message = data.get("message")
    results = data.get("results")
    return SecIDResponse(
        secid_query=query if isinstance(query, str) else secid,
        status=status if isinstance(status, str) and status else "error",
        results=[r for r in results if isinstance(r, dict)] if isinstance(results, list) else [],
        message=message if isinstance(message, str) else None,
    )


class SecIDClient:
    """HTTP client for the SecID resolve API.

    Args:
        base_url: API base URL. Defaults to the public SecID service.
    """

    def __init__(self, base_url: str = DEFAULT_BASE_URL, timeout: int = DEFAULT_TIMEOUT) -> None:
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout

    def resolve(self, secid: str) -> SecIDResponse:
        """Resolve a SecID string to URL(s).

        The # character is automatically encoded as %23 in the query parameter.

        Args:
            secid: Full SecID string, e.g. "secid:advisory/mitre.org/cve#CVE-2021-44228"

        Returns:
            SecIDResponse with status, results, and optional message. Never
            raises for network, HTTP, or response-format problems: those come
            back as status="error" with an explanatory message.
        """
        def error(message: str) -> SecIDResponse:
            return SecIDResponse(secid_query=secid, status="error", message=message)

        try:
            encoded = urllib.parse.quote(secid, safe="")
        except UnicodeEncodeError:  # a lone surrogate cannot be UTF-8 encoded
            return error("SecID is not valid Unicode (unpaired surrogate)")
        url = f"{self.base_url}/api/v1/resolve?secid={encoded}"

        # Every failure mode becomes status="error"; resolve() never raises.
        # OSError covers URLError, socket timeouts, and connection resets;
        # HTTPException covers malformed HTTP (IncompleteRead, BadStatusLine);
        # ValueError covers an unusable base_url ("unknown url type").
        try:
            req = urllib.request.Request(url, headers={
                "Accept": "application/json",
                "User-Agent": "secid-python-client/1.0",
            })
            with urllib.request.urlopen(req, timeout=self.timeout) as resp:
                body = resp.read(MAX_RESPONSE_BYTES + 1)
        except urllib.error.HTTPError as e:
            # Non-2xx: the resolver may still send a JSON envelope.
            try:
                body = e.read(MAX_RESPONSE_BYTES + 1)
            except (OSError, http.client.HTTPException, ValueError) as read_err:
                return error(f"HTTP {e.code}: error reading body: {read_err}")
            data = _parse_json(body)
            if not isinstance(data, dict):
                text = body[:200].decode("utf-8", errors="replace")
                return error(f"HTTP {e.code}: {text}")
            return _envelope(data, secid)
        except urllib.error.URLError as e:
            if isinstance(e.reason, _TIMEOUTS):
                return error(f"Request timeout after {self.timeout}s: {e.reason}")
            return error(f"Connection error: {e.reason}")
        except _TIMEOUTS as e:
            # A timeout while waiting for or reading the response is raised
            # bare, not wrapped in URLError.
            return error(f"Request timeout after {self.timeout}s: {e}")
        except (OSError, http.client.HTTPException) as e:
            return error(f"Connection error: {e}")
        except ValueError as e:
            return error(f"Invalid request URL {url!r}: {e}")

        if len(body) > MAX_RESPONSE_BYTES:
            return error(f"Response exceeds {MAX_RESPONSE_BYTES} byte limit")
        data = _parse_json(body)
        if data is _INVALID:
            return error("Invalid response (not JSON)")
        if not isinstance(data, dict):
            return error(f"Invalid response: expected a JSON object, got {type(data).__name__}")
        return _envelope(data, secid)

    def best_url(self, secid: str) -> str | None:
        """Resolve a SecID and return the highest-weight URL, or None."""
        return self.resolve(secid).best_url

    def lookup(self, type: str, identifier: str) -> SecIDResponse:
        """Cross-source search: find an identifier across all sources of a type.

        Equivalent to resolve(f"secid:{type}/{identifier}").

        Args:
            type: SecID type (advisory, weakness, ttp, control, capability, methodology, disclosure, regulation, entity, reference).
            identifier: The identifier to search for, e.g. "CVE-2021-44228".
        """
        return self.resolve(f"secid:{type}/{identifier}")


def main() -> None:
    if len(sys.argv) < 2 or sys.argv[1] in ("-h", "--help"):
        print("Usage: secid [--json] <secid>")
        print()
        print("Examples:")
        print('  secid "secid:advisory/mitre.org/cve#CVE-2021-44228"')
        print('  secid --json "secid:advisory/mitre.org/cve#CVE-2021-44228"')
        print('  secid "secid:advisory/CVE-2021-44228"')
        sys.exit(0)

    json_mode = "--json" in sys.argv
    args = [a for a in sys.argv[1:] if a != "--json"]
    if not args:
        print("Error: no SecID provided", file=sys.stderr)
        sys.exit(1)

    client = SecIDClient()
    response = client.resolve(args[0])

    if json_mode:
        print(json.dumps({
            "secid_query": response.secid_query,
            "status": response.status,
            "results": response.results,
            "message": response.message,
        }, indent=2))
    elif response.status in ("found", "corrected"):
        url = response.best_url
        if url:
            if response.was_corrected:
                print(f"(corrected to: {_sanitize_terminal(str(response.results[0].get('secid', '')))})", file=sys.stderr)
            print(_sanitize_terminal(url))
        else:
            for r in response.registry_results:
                print(json.dumps(r, indent=2))
    elif response.status == "related":
        for r in response.results:
            print(json.dumps(r, indent=2))
    else:
        msg = response.message or "No results"
        print(f"{response.status}: {_sanitize_terminal(msg)}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: typescript/src/secid-client.ts
/**
 * SecID client — resolve security identifiers to URLs.
 *
 * Zero runtime dependencies (fetch only).
 *
 * SecID is a universal grammar for security knowledge:
 *     secid:type/namespace/name[@version]#subpath
 *
 * API: GET https://secid.cloudsecurityalliance.org/api/v1/resolve?secid={encoded}
 *
 * IMPORTANT: The # character in SecID strings must be encoded as %23 in the
 * URL query parameter. This is the #1 failure mode for new clients.
 *
 * Usage:
 *     import { SecIDClient } from "@cloudsecurityalliance/secid";
 *     const client = new SecIDClient();
 *     const response = await client.resolve("secid:advisory/mitre.org/cve#CVE-2021-44228");
 *     console.log(response.bestUrl);
 */

export const VERSION = "0.1.0";

const DEFAULT_BASE_URL = "https://secid.cloudsecurityalliance.org";
const DEFAULT_TIMEOUT_MS = 30_000; // 30 seconds
const MAX_RESPONSE_BYTES = 10 * 1024 * 1024; // 10 MB

async function readBodyWithByteLimit(resp: Response, maxBytes: number): Promise<string> {
  const stream = resp.body;
  if (!stream) return "";

  const reader = stream.getReader();
  const chunks: Uint8Array[] = [];
  let totalBytes = 0;

  while (true) {
    const { done, value } = await reader.read();
    if (done) break;
    if (!value) continue;

    totalBytes += value.byteLength;
    if (totalBytes > maxBytes) {
      try {
        await reader.cancel();
      } catch {
        // Best-effort cancel; limit error still propagates below.
      }
      throw new Error(`Response exceeds ${maxBytes} byte limit`);
    }
    chunks.push(value);
  }

  const merged = new Uint8Array(totalBytes);
  let offset = 0;
  for (const chunk of chunks) {
    merged.set(chunk, offset);
    offset += chunk.byteLength;
  }
  return new TextDecoder().decode(merged);
}

/** A single result that resolved to a URL. */
export interface ResolutionResult {
  secid: string;
  weight: number;
  url: string;
  content_type?: string;        // MIME type (e.g., "text/html", "application/json")
  parsability?: string;          // "structured" or "scraped"
  schema?: string;               // SecID reference to data schema
  parsing_instructions?: string; // SecID reference to parsing instruction doc
  auth?: string;                 // Free-text auth description
}

/** A single result containing registry/browsing data. */
export interface RegistryResult {
  secid: string;
  data: Record<string, unknown>;
}

/** Raw result from the API — either resolution or registry type. */
export type SecIDResult = ResolutionResult | RegistryResult;

/** Response from the SecID resolve API. */
export class SecIDResponse {
  /** The query string echoed back (decoded form). */
  readonly secidQuery: string;
  /** One of: found, corrected, related, not_found, error. */
  readonly status: string;
  /** Result objects — either resolution or registry type. */
  readonly results: SecIDResult[];
  /** Guidance text on not_found/error, undefined otherwise. */
  readonly message?: string;

  constructor(data: {
    secid_query: string;
    status: string;
    results: SecIDResult[];
    message?: string | null;
  }) {
    this.secidQuery = data.secid_query;
    this.status = data.status;
    // Keep only object entries: the getters below read properties off each
    // result, and a hostile resolver can put anything in the array.
    this.results = Array.isArray(data.results) ? data.results.filter(isObject) : [];
    this.message = data.message ?? undefined;
  }

  /** Highest-weight URL from resolution results, or undefined. */
  get bestUrl(): string | undefined {
    const resolved = this.resolutionResults;
    return resolved.length > 0 ? validateUrl(resolved[0].url) : undefined;
  }

  /** True if the server corrected the input. */
  get wasCorrected(): boolean {
    return this.status === "corrected";
  }

  /**
   * Only results with a numeric weight and a string url, sorted by weight
   * descending. A missing, null, string, or non-finite weight is not a
   * resolution result (sorting it as 0 or NaN would corrupt the order).
   */
  get resolutionResults(): ResolutionResult[] {
    return this.results
      .filter((r): r is ResolutionResult => {
        const w = (r as { weight?: unknown }).weight;
        return typeof w === "number" && Number.isFinite(w)
          && typeof (r as { url?: unknown }).url === "string";
      })
      .sort((a, b) => b.weight - a.weight);
  }

  /** Only results with data (registry/browsing info). */
  get registryResults(): RegistryResult[] {
    return this.results
      .filter((r): r is RegistryResult => "data" in r);
  }
}

/** True for a non-null, non-array object. */
function isObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

/**
 * Build a SecIDResponse from an untrusted parsed JSON object, type-checking
 * each field. Wrong-typed fields fall back to defaults.
 */
function envelope(data: Record<string, unknown>, secid: string): SecIDResponse {
  const { secid_query, status, results, message } = data;
  return new SecIDResponse({
    secid_query: typeof secid_query === "string" ? secid_query : secid,
    status: typeof status === "string" && status !== "" ? status : "error",
    results: Array.isArray(results) ? (results.filter(isObject) as unknown as SecIDResult[]) : [],
    message: typeof message === "string" ? message : undefined,
  });
}

// The resolver response is untrusted (a hostile, federated, or MITM'd resolver
// is in scope). Only http(s) URLs may be surfaced as a "best URL"; anything
// else (javascript:, data:, file:, ...) or a relative URL is rejected.
const ALLOWED_URL_SCHEMES = new Set(["https:", "http:"]);

/** Return url only if it is an absolute http(s) URL, else undefined. */
function validateUrl(url: unknown): string | undefined {
  if (typeof url !== "string" || url.length === 0) return undefined;
  let parsed: URL;
  try {
    parsed = new URL(url);
  } catch {
    return undefined;
  }
  return ALLOWED_URL_SCHEMES.has(parsed.protocol) ? url : undefined;
}

/** HTTP client for the SecID resolve API. */
export class SecIDClient {
  private readonly baseUrl: string;
  private readonly timeoutMs: number;

  /**
   * @param baseUrl - API base URL. Defaults to the public SecID service.
   * @param timeoutMs - Request timeout in milliseconds. Defaults to 30 seconds.
   */
  constructor(baseUrl: string = DEFAULT_BASE_URL, timeoutMs: number = DEFAULT_TIMEOUT_MS) {
    this.baseUrl = baseUrl.replace(/\/+$/, "");
    this.timeoutMs = timeoutMs;
  }

  /**
   * Resolve a SecID string to URL(s).
   *
   * The # character is automatically encoded as %23 in the query parameter.
   *
   * @param secid - Full SecID string, e.g. "secid:advisory/mitre.org/cve#CVE-2021-44228"
   * @returns Never rejects for network, HTTP, or response-format problems:
   *   those come back as status "error" with an explanatory message.
   */
  async resolve(secid: string): Promise<SecIDResponse> {
    const encoded = encodeURIComponent(secid);
    const url = `${this.baseUrl}/api/v1/resolve?secid=${encoded}`;

    const error = (message: string) =>
      new SecIDResponse({ secid_query: secid, status: "error", results: [], message });

    // Transport: fetch + bounded body read. Nothing in resolve() throws —
    // every failure becomes status="error".
    let body: string;
    let httpStatus: number;
    let ok: boolean;
    try {
      const resp = await fetch(url, {
        headers: {
          Accept: "application/json",
          "User-Agent": "secid-typescript-client/1.0",
        },
        signal: AbortSignal.timeout(this.timeoutMs),
      });
      httpStatus = resp.status;
      ok = resp.ok;
      const contentLength = resp.headers.get("content-length");
      if (contentLength && parseInt(contentLength, 10) > MAX_RESPONSE_BYTES) {
        try {
          await resp.body?.cancel();
        } catch {
          // Best-effort; the size error is what matters.
        }
        return error(`Response exceeds ${MAX_RESPONSE_BYTES} byte limit`);
      }
      body = await readBodyWithByteLimit(resp, MAX_RESPONSE_BYTES);
    } catch (err) {
      const msg = err instanceof Error ? err.message : String(err);
      if (msg.startsWith("Response exceeds ")) return error(msg);
      // AbortSignal.timeout rejects with a TimeoutError DOMException, both
      // while waiting for headers and while streaming the body.
      const name = (err as { name?: unknown } | null)?.name;
      if (name === "TimeoutError" || name === "AbortError") {
        return error(`Request timeout after ${this.timeoutMs}ms: ${msg}`);
      }
      return error(`Connection error: ${msg}`);
    }

    // Parsing: a separate step, so a bad body is not reported as a
    // connection problem.
    let data: unknown;
    try {
      data = JSON.parse(body);
    } catch (err) {
      if (!ok) return error(`HTTP ${httpStatus}: ${body.slice(0, 200)}`);
      const msg = err instanceof Error ? err.message : String(err);
      return error(`Invalid response (not JSON): ${msg}`);
    }
    if (!isObject(data)) {
      if (!ok) return error(`HTTP ${httpStatus}: ${body.slice(0, 200)}`);
      const kind = data === null ? "null" : Array.isArray(data) ? "array" : typeof data;
      return error(`Invalid response: expected a JSON object, got ${kind}`);
    }
    return envelope(data, secid);
  }

  /**
   * Resolve a SecID and return the highest-weight URL, or undefined.
   */
  async bestUrl(secid: string): Promise<string | undefined> {
    return (await this.resolve(secid)).bestUrl;
  }

  /**
   * Cross-source search: find an identifier across all sources of a type.
   *
   * Equivalent to resolve(`secid:${type}/${identifier}`).
   *
   * @param type - SecID type (advisory, weakness, ttp, control, capability, methodology, disclosure, regulation, entity, reference).
   * @param identifier - The identifier to search for, e.g. "CVE-2021-44228".
   */
  async lookup(type: string, identifier: string): Promise<SecIDResponse> {
    return this.resolve(`secid:${type}/${identifier}`);
  }
}
